NVDAObjects/UIA: recognize UWP tooltips

josephsl opened the pull request:
### Link to issue number:
Fixes #8118 

### Summary of the issue:
Recognize UWP/XAML tooltips such as ones from Microsoft Edge as proper tooltips.

### Description of how this pull request fixes the issue:
Tooltips from universal apps are powered by XAML and has a specific class name (for now). Prior to this, these tooltips were not announced by NVDA. Cases include Edge's Settings button tooltip, which displays hotkey information (Alt+X) in Insider Preview builds. Thus recognize them as proper tooltips through an overlay class powered by NVDAObjects.behaviors.ToolTip.

### Testing performed:
Tested Edge on Windows 10 Versions 1703 (Creators Update), 1709 (Fall Creators Update) and 1803 preview (Spring Creators Update).

### Known issues with pull request:
None at this time.

### Change log entry:
Bug fixes: In Windows 10, NVDA will announce tooltips from universal apps if NVDA is configured to report tooltips in object presentation dialog.

josephsl committed:
NVDAObjects/UIA: recognize UWP tooltips. Re #8118.

Tooltips from universal apps are powered by XAML and has a specific class name (for now). Thus recognize them as proper tool tips.

leonardder commented:
[**@josephsl**](https://github.com/josephsl) commented on [25 mrt. 2018 19:44 CEST](https://github.com/nvaccess/nvda/pull/8124#issue-177301490 "2018-03-25T17:44:03Z - Replied by Github Reply Comments"):
> Tooltips from universal apps are powered by XAML and has a specific class name (for now).

What UIA control type do these have? It would be nice if we could also support WPF tooltips this way, though I have no idea where to find these for testing purposes.

josephsl commented:
Hi, haven’t investigated that far, but I assume they’re regular tooltips.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Sunday, March 25, 2018 11:43 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] NVDAObjects/UIA: recognize UWP tooltips (#8124)

 

 <https://github.com/josephsl> @josephsl commented on 25 mrt. 2018 19:44 CEST <https://github.com/nvaccess/nvda/pull/8124#issue-177301490> :

Tooltips from universal apps are powered by XAML and has a specific class name (for now).

What UIA control type do these have? It would be nice if we could also support WPF tooltips this way, though I have no idea where to find these for testing purposes.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8124#issuecomment-376062169> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkGyBpX5NSXPPdd8HkEUrYTteTj-3ks5tiI33gaJpZM4S6OxY> .

leonardder commented:
[**@josephsl**](https://github.com/josephsl) commented on [26 mrt. 2018 09:21 CEST](https://github.com/nvaccess/nvda/pull/8124#issuecomment-376069890 "2018-03-26T07:21:49Z - Replied by Github Reply Comments"):
> Hi,
> 
> The tooltips are recognized as tooltips. Thanks.

In that case, why not just treat every UIA tooltip as tooltip?

josephsl commented:
Hi, not yet, as one of them is volume slider tooltip, which is annoying. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Monday, March 26, 2018 1:11 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] NVDAObjects/UIA: recognize UWP tooltips (#8124)

 

 <https://github.com/josephsl> @josephsl commented on 26 mrt. 2018 09:21 CEST <https://github.com/nvaccess/nvda/pull/8124#issuecomment-376069890> :

Hi,

The tooltips are recognized as tooltips. Thanks.

In that case, why not just treat every UIA tooltip as tooltip?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8124#issuecomment-376081679> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkMTggz3o6NaZowfd-4OyIQshA_T7ks5tiKKtgaJpZM4S6OxY> .

leonardder commented:
[**@josephsl**](https://github.com/josephsl) commented on [26 mrt. 2018 16:48 CEST](https://github.com/nvaccess/nvda/pull/8124#issuecomment-376192606 "2018-03-26T14:48:28Z - Replied by Github Reply Comments"):
> Hi, not yet, as one of them is volume slider tooltip, which is annoying. Thanks.

Could you please elaborate?

josephsl commented:
Hi, when you change audio volume, a tooltip comes up and fires tooltip open event. Although the framework is different, it is still annoying. Thus, I’ve intentionally limited tooltip announcement to XAML tooltips for now. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Monday, March 26, 2018 8:09 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] NVDAObjects/UIA: recognize UWP tooltips (#8124)

 

 <https://github.com/josephsl> @josephsl commented on 26 mrt. 2018 16:48 CEST <https://github.com/nvaccess/nvda/pull/8124#issuecomment-376192606> :

Hi, not yet, as one of them is volume slider tooltip, which is annoying. Thanks.

Could you please elaborate?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8124#issuecomment-376200043> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkC34-T09FRHEiqWNAs7WBxYira2rks5tiQR8gaJpZM4S6OxY> .

josephsl committed:
NVDAObjects/UIA: rename UWPToolTip to XAMLToolTip to better reflect where this is coming from. Re #8118

leonardder commented:
[**@josephsl**](https://github.com/josephsl) commented on [26 Mar 2018, 17:44 CEST](https://github.com/nvaccess/nvda/pull/8124#issuecomment-376212903 "2018-03-26T15:44:31Z - Replied by Github Reply Comments"):
> Hi, when you change audio volume, a tooltip comes up and fires tooltip open event. Although the framework is different, it is still annoying.

I think many people consider tool tips to be annoying. I prefer consistency over subjective decisions about what is considered annoying and what is not. @MichaeldCurran: Thoughts?

michaelDCurran commented:
I tend to agree with @leonardder.  If a choose to  have tooltips reported, then I want to hear all tooltips. Obviously this can get annoying, thus why it is disabled by default in NVDA.

josephsl commented:
Hi,

Testing shows that I had incomplete assumptions: tooltip for volume slider is fired infrequently, thus I'll follow above recommendations.

Thanks.

josephsl committed:
NVDAObjects/UIA: relax tool tip announcement requirement by supporting ones from volume slider and other controls. Re #8118.

Commented by Mick Curran (NV Access) and others: wrong assumption from my part - volume slider does not raise tool tip open event frequently, so this is safe to be called out. This is done by just checking for control type (same one as Windows 8.x toasts).

josephsl commented:
Hi,

Note that new UIA tool tip handling works properly on Windows 8 and later. An additional benefit is that NVDA will now also announce volume changes (if that is announced, that is).

Thanks.

Robert-J-H commented:
Hi
I must say that the Volume Level tooltip is very distracting.
Although it closes essentially #5726, the opinions uttered there aren't by no way unanimous.

It is especially annoying if you adjust the volume during say all.
It might happen that it is announced much later (after e.g. a paragraph).

It is not very user-friendly if the possible decision for tool tips is
"Hav'em all or none".
It is a special case tool tip because:
1. It gives the immediate result for a user input 
2. The effect can also be heard by other means (playing audio, running speech)
As a musician, I clearly prefer the second way.
More so as the user intentionally lowers or rises the volume because of what she/he heard or didn't hear and not because the numeric value wasn't to her/his liking.
In other words, the cause/effect chain is broken or at least forked into two outputs. 
It would be great to have this tool tip only announced for when the audio device is idle.
Does this make sense? ;-)

leonardder commented:
[**@Robert-J-H**](https://github.com/Robert-J-H) commented on [21 Apr 2018, 08:45 CEST](https://github.com/nvaccess/nvda/pull/8124#issuecomment-383272454 "2018-04-21T06:45:03Z - Replied by Github Reply Comments"):
> It is especially annoying if you adjust the volume during say all.
> It might happen that it is announced much later (after e.g. a paragraph).

May be this is more a general problem of reporting tooltips during say all, not specifically to the volume tool tip.

> It is not very user-friendly if the possible decision for tool tips is
> "Hav'em all or none".

What do you suggest otherwise? I don't think it is user friendly either to distinguish between several tool tips the user does not know about.

josephsl commented:
Hi, this is due to the fact that the object in question fires tool tip opened event infrequently. I had similar objections due to this being not quite consistent, but ultimately it is up to users to decide. Thanks.

 

From: Robert Haenggi <notifications@github.com> 
Sent: Saturday, April 21, 2018 1:45 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] NVDAObjects/UIA: recognize UWP tooltips (#8124)

 

Hi
I must say that the Volume Level tooltip is very distracting.
Although it closes essentially #5726 <https://github.com/nvaccess/nvda/issues/5726> , the opinions uttered there aren't by no way unanimous.

It is especially annoying if you adjust the volume during say all.
It might happen that it is announced much later (after e.g. a paragraph).

It is not very user-friendly if the possible decision for tool tips is
"Hav'em all or none".
It is a special case tool tip because:

1.	It gives the immediate result for a user input
2.	The effect can also be heard by other means (playing audio, running speech)
As a musician, I clearly prefer the second way.
More so as the user intentionally lowers or rises the volume because of what she/he heard or didn't hear and not because the numeric value wasn't to her/his liking.
In other words, the cause/effect chain is broken or at least forked into two outputs.
It would be great to have this tool tip only announced for when the audio device is idle.
Does this make sense? ;-)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8124#issuecomment-383272454> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkJXbJI2aG3PnmddXNY7fordSwqctks5tqtVwgaJpZM4S6OxY> .

Robert-J-H commented:
Yes, the user should be able to decide but the choices are like those of a political election with only one candidate.
Over the last years, the notion of tooltips has somewhat changed.  
Let me quote the doc string of the tooltip class for the original definition:  

> Provides information about an item over which the user is hovering a cursor.
> The object should fire a show event when it appears.

MSDN gives a similar description, perhaps even narrower as they refer mainly to the mouse pointer.  
However, one has to read between the lines to establish a different view point.
In contrast to the classic definition, tooltips can now appear outside the app.  
In UIA terms, this would be "Peripheral", I think.  
Let's call this new type flyouts.
Some classic tooltips are quite useful, e.g. those that you get in an explorer window with modification date, size etc. for a file.
Taskbar tooltips that are announced whilst the focus isn't there are rather a nuisance IMHO.
It is perhaps difficult to implement a finer distinction in NVDA core (e.g. by adding flyouts to
the object presentation dialog).

I think the best we can do is adding filter capability to the tooltip class such that add-ons can profit from it.

This is the perfect occasion to use the extensionPoints.Filter class.
After all, it has been introduced for exactly that purpose, has it not.

josephsl commented:
Hi,

Any additional comments?

The filter mechanism is useful for text processing, not really a way to let overlay classes change things here and there. To do that, one must be willing to add conditions in the class itself, or let global plugins and app modules take ownership of this class (inheritance and overriding).

Thanks.

Robert-J-H commented:
I haven't the extensionPoints code at my fingertips right now but I
think there's also the Decider class.
Registered functions would be notified about the tooltip source and
the text to be spoken.
They can then return either True or False to allow or prevent the
speaking of it. Thx

On 17/06/2018, Joseph Lee <notifications@github.com> wrote:
> Hi,
>
> Any additional comments?
>
> The filter mechanism is useful for text processing, not really a way to let
> overlay classes change things here and there. To do that, one must be
> willing to add conditions in the class itself, or let global plugins and app
> modules take ownership of this class (inheritance and overriding).
>
> Thanks.
>
> --
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/nvaccess/nvda/pull/8124#issuecomment-397848565

gregjozk commented:
Hello,

what is a status of this pr?

thx.

regards,
Jožef

josephsl commented:
Hi,

Regarding deciders: this means extension points must be willing to deal with cases where tooltips are not announced because someone said "no" while the decider was running. I think this might not be what we want.

For now, I'd treat tooltips from UIA as just tooltips. If there are no major objections, I propose that this be ready for master branch soon.

Thanks.

ehollig commented:
Hello, What is the status of this PR? It looks like it has been incubating for a while now.

josephsl commented:
Hi, some small things here and there, and apart from that, and once it is approved, it should be ready for master branch. In case review comments are posted for this one, I’ll address them in November. Thanks.

 

From: Ethan Holliger <notifications@github.com> 
Sent: Friday, October 12, 2018 5:20 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] NVDAObjects/UIA: recognize UWP tooltips (#8124)

 

Hello, What is the status of this PR? It looks like it has been incubating for a while now.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8124#issuecomment-429494948> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkFUHZ3AQk3-gV5cANIQZvSDAgTmyks5ukTG5gaJpZM4S6OxY> .

josephsl commented:
Hi,

Actually, it is ready to go - I think it's just the matter of when it'll be included in master branch, and I can address review comments in October.

Thanks.

leonardder commented:
I removed the incubating label from this as we're no longer using that label. Having said that, the fact that this has been incubating for a while suggest that the code is of enough quality and has been approved in the past.

josephsl commented:
Hi, just a few things here and there and should be ready for another round of reviews – at least I heard no major reports about this in Windows 10 App Essentials, so it could serve as a basis for a review next week. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Saturday, May 18, 2019 9:44 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] NVDAObjects/UIA: recognize UWP tooltips (#8124)

 

I removed the incubating label from this as we're no longer using that label. Having said that, the fact that this has been incubating for a while suggest that the code is of enough quality and has been approved in the past.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8124?email_source=notifications&email_token=AB4AXEEK2HEVJ2HMTZBF5YLPWAW5DA5CNFSM4EXI5RMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVWR5AA#issuecomment-493690496> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEDU2FFVQUCDT6JVZRLPWAW5DANCNFSM4EXI5RMA> .

josephsl commented:
Hi,

Let me merge in master first.

Thanks.

josephsl commented:
Hi,

Regarding review bits: yes, approved, but it won't hurt to let another pair of eyes/ears/hands take a look at this one before making this out to master.

Thanks.

Robert-J-H commented:
Hm, I still see a lot of issues (see my remarks above).
The tooltip behaviour is somewhat erratic (not necessarily related to
UIA alone). For instance:
The tool tip for the last focused task bar icon is often announced
between focus changes within an application. In general, this doesn't
produce anything valuable.
Tool tips for desktop icons with focus (e.g. shortcut target) are
often wrong as they can come from the mouse object, rather than the
focused object.
I see this behaviour also in Reaper.

On 18/05/2019, Joseph Lee <notifications@github.com> wrote:
> Hi,
>
> Regarding review bits: yes, approved, but it won't hurt to let another pair
> of eyes/ears/hands take a look at this one before making this out to
> master.
>
> Thanks.
>
> --
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/nvaccess/nvda/pull/8124#issuecomment-493691440

dingpengyu commented:
This PR has the following error in the build:
Failure: AppVeyor build failed
Hope that can be fixed
thank
cc
@josephsl

josephsl commented:
Hi. Likely caused by the old code, which will be fixed once master branch is merged. Thanks.

 

From: eric <notifications@github.com> 
Sent: Tuesday, May 21, 2019 6:10 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] NVDAObjects/UIA: recognize UWP tooltips (#8124)

 

This PR has the following error in the build:
Failure: AppVeyor build failed
Hope that can be fixed
thank
cc
@josephsl <https://github.com/josephsl> 

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8124?email_source=notifications&email_token=AB4AXEB2IAZNOO24T6EDMF3PWPYC5A5CNFSM4EXI5RMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV33HPA#issuecomment-494384060> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXECWLVRL2YIIS3H6THLPWPYC5ANCNFSM4EXI5RMA> .

AmadeusW commented:
How can I test changes in this PR? @josephsl do you have files I can use to patch NVDA 2019.1.1? I can install any other version, too, if that would make it easier for you to share the fix.

I'd like to verify whether this addresses issue where XAML tooltips are not announced in Visual Studio (VS is a Win32 app using WPF\XAML, not an Universal Windows app)

josephsl commented:
Hi, I have suspended development of this PR for now. Thanks.

 

From: Amadeusz Wieczorek <notifications@github.com> 
Sent: Friday, June 14, 2019 12:47 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] NVDAObjects/UIA: recognize UWP tooltips (#8124)

 

How can I test changes in this PR? @josephsl <https://github.com/josephsl>  do you have files I can use to patch NVDA 2019.1.1? I can install any other version, too, if that would make it easier for you to share the fix.

I'd like to verify whether this addresses issue where XAML tooltips are not announced in Visual Studio (VS is a Win32 app using WPF\XAML, not an Universal Windows app)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8124?email_source=notifications&email_token=AB4AXEAZKAN4TZ6TG2HMSBDP2PYSXA5CNFSM4EXI5RMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXXZHQY#issuecomment-502240195> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXECCAV4EIOAZLR55KNTP2PYSXANCNFSM4EXI5RMA> .

AmadeusW commented:
that's fine, I realized that NVDA does read some tooltips in VS, so this PR wouldn't have helped my case. Thank you!

josephsl commented:
Hi,

I'll make sure this PR passes checks.

Thanks.

josephsl commented:
Hi, I’ll rebase it on master unless there is no conflict merging this into master. Thanks.

 

From: Reef Turner <notifications@github.com> 
Sent: Friday, July 26, 2019 6:42 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] NVDAObjects/UIA: recognize UWP tooltips (#8124)

 

@feerrenrut approved this pull request.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8124?email_source=notifications&email_token=AB4AXEBHZQEFJ53KQ7R7QB3QBL5JTA5CNFSM4EXI5RMKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB7WQYMY#pullrequestreview-267193395> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEDBMX235UT4ZYPD7E3QBL5JTANCNFSM4EXI5RMA> .
